File: individual/src/main/java/requests/ErrorPacket.java
package requests;

import java.nio.ByteBuffer;

import static parameters.OpCode.ERROR;

public class ErrorPacket extends Packet {

    private final String message;
    private final short error;

    public ErrorPacket(short error, String message) {
        super(ERROR);
        this.error = error;
        this.message = message;
    }

    @Override
    public int size() {
        return Character.BYTES + Character.BYTES + message.length() + 1;
    }

    public String getMessage() {
        return message;
    }

    @Override
    public byte[] serialize() {
        ByteBuffer buf = ByteBuffer.allocate(size());
        buf.putShort(getOpCode());
        buf.putShort(error);
        buf.put(message.getBytes());
        buf.put((byte) 0);
        return buf.array();
    }
}

File: udp/src/main/kotlin/ru/hse/anstkras/tftp/ErrorPacket.kt
package ru.hse.anstkras.tftp

import java.nio.ByteBuffer

class ErrorPacket(private val errorCode: Int, private val errorMessage: ByteBuffer) : Packet {

    //    2 bytes     2 bytes      string    1 byte
    //    -----------------------------------------
    //    | Opcode |  ErrorCode |   ErrMsg   |   0  |
    //    -----------------------------------------
    override fun getBytesRepresentation(): ByteBuffer {
        val buffer = ByteBuffer.allocate(errorMessage.remaining() + 5)
        buffer.putShort(5)
        buffer.putShort(errorCode.toShort())
        buffer.put(errorMessage)
        buffer.put(0)
        return buffer
    }
}